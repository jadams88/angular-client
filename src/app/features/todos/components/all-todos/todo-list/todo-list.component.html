<mat-card (click)="selected.emit(todo)">
  <mat-card-subtitle>{{ todo.title }}</mat-card-subtitle>
  <mat-card-content fxLayout="row" fxLayoutAlign="space-between center">
    <span>{{ todo.description }}</span>
    <mat-slide-toggle
      [checked]="todo.completed"
      (click)="$event.stopPropagation()"
      (change)="updateTodoCompletedStatus(todo, $event.checked)"
      >Completed</mat-slide-toggle
    >
  </mat-card-content>
  <mat-card-actions fxLayoutAlign="end">
    <mat-icon class="button" (click)="delete.emit(todo); $event.stopPropagation()">delete</mat-icon>
  </mat-card-actions>
</mat-card>
