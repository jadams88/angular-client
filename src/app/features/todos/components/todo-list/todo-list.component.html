<mat-card *ngFor="let todo of todos" (click)="selected.emit(todo.id)">
  <mat-card-subtitle>{{ todo.title }}</mat-card-subtitle>
  <mat-card-content>{{ todo.description }}</mat-card-content>
  <mat-card-actions>
    <mat-icon class="button" (click)="delete.emit(todo); $event.stopPropagation()">delete</mat-icon>
  </mat-card-actions>
</mat-card>
