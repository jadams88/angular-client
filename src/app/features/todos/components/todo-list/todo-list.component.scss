.mat-card {
  margin-bottom: 1em;
  :last-of-type {
    margin-bottom: none;
  }
  .button {
    cursor: pointer;
  }
}
