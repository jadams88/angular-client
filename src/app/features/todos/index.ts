export { Todo } from './todos.model';
export { TodoModule } from './todos.module';
export { TodoService } from './services/todos.service';
export * from './actions/todos.actions';
