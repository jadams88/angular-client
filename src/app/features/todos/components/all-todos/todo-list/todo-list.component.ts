import { Component, Input, ChangeDetectionStrategy, Output, EventEmitter } from '@angular/core';
import { Todo } from '@app/features/todos';

@Component({
  selector: 'demo-todo-list',
  templateUrl: './todo-list.component.html',
  styleUrls: ['./todo-list.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class TodoListComponent {
  @Input()
  todo: Todo;
  @Output()
  selected = new EventEmitter<Todo>();
  @Output()
  delete = new EventEmitter<Todo>();
  @Output()
  update = new EventEmitter<Todo>();

  public updateTodoCompletedStatus(todo: Todo, completed: boolean) {
    const updatedTodo: Todo = { ...todo, completed };
    this.update.emit(updatedTodo);
  }
}
