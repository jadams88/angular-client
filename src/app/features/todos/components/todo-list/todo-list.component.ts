import { Component, Input, ChangeDetectionStrategy, Output, EventEmitter } from '@angular/core';
import { Todo } from '@app/features/todos';

@Component({
  selector: 'demo-todo-list',
  templateUrl: './todo-list.component.html',
  styleUrls: ['./todo-list.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class TodoListComponent {
  @Input()
  todos: Todo[];
  @Output()
  selected = new EventEmitter<string>();
  @Output()
  delete = new EventEmitter<Todo>();
}
