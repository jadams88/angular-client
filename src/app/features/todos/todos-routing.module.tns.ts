import { NgModule } from '@angular/core';
import { Routes } from '@angular/router';
import { TodosComponent } from './components/todos.component';
import { NativeScriptRouterModule } from 'nativescript-angular/router';

const routes: Routes = [
  { path: '/:id', component: TodosComponent },
  { path: '', pathMatch: 'full', component: TodosComponent }
];

@NgModule({
  imports: [NativeScriptRouterModule.forChild(routes)],
  exports: [NativeScriptRouterModule]
})
export class TodoRoutingModule {}
