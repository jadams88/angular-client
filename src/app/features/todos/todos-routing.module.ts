import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { TodosComponent } from './components/todos.component';

const routes: Routes = [{ path: '', pathMatch: 'full', component: TodosComponent }];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class TodoRoutingModule {}
