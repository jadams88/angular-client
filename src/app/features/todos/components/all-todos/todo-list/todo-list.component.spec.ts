import { async, ComponentFixture, TestBed, fakeAsync, tick } from '@angular/core/testing';

import { TodoListComponent } from './todo-list.component';
import { NO_ERRORS_SCHEMA, DebugElement, Component } from '@angular/core';
import { Todo } from '@app/features/todos';
import { By } from '@angular/platform-browser';

@Component({
  template: `
    <demo-todo-list [todo]="todo"> </demo-todo-list>
  `
})
export class TestParentComponent {
  todo: Todo;
}

describe('TodoListComponent ', () => {
  let parentComponent: TestParentComponent;
  let parentFixture: ComponentFixture<TestParentComponent>;
  let component: TodoListComponent;
  let fixture: ComponentFixture<TodoListComponent>;
  let debugEl: DebugElement;
  let nativeEl: HTMLElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [TestParentComponent, TodoListComponent],
      schemas: [NO_ERRORS_SCHEMA]
    }).compileComponents();
  }));

  beforeEach(() => {
    parentFixture = TestBed.createComponent(TestParentComponent);
    parentComponent = parentFixture.componentInstance;
    fixture = TestBed.createComponent(TodoListComponent);
    component = fixture.componentInstance;
    debugEl = fixture.debugElement;
    nativeEl = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('todo', () => {
    it('should displays the todo details when set', () => {
      const todo: Todo = {
        id: '1',
        user: '1',
        title: 'some title',
        description: 'some description',
        completed: true
      };

      component.todo = todo;
      fixture.detectChanges();

      expect(nativeEl).toMatchSnapshot();
    });
  });

  describe('selected', () => {
    it('should be raised with current todo when the card is clicked', fakeAsync(() => {
      const todo: Todo = {
        id: '1',
        user: '1',
        title: 'some title',
        description: 'some description',
        completed: true
      };

      component.todo = todo;
      fixture.detectChanges();

      let selectedTodo: Todo;
      component.selected.subscribe(event => {
        selectedTodo = event;
      });

      const matCardList = debugEl.queryAll(By.css('mat-card'));
      matCardList[0].triggerEventHandler('click', null);

      tick();

      expect(selectedTodo).toEqual(todo);
    }));
  });

  describe('delete', () => {
    it('should be raised with the current todo when the delete icon is clicked', fakeAsync(() => {
      const todo: Todo = {
        id: '1',
        user: '1',
        title: 'some title',
        description: 'some description',
        completed: true
      };

      component.todo = todo;
      fixture.detectChanges();

      let deletedTodo: Todo;
      component.delete.subscribe(event => {
        deletedTodo = event;
      });

      const deleteButtons = debugEl.queryAll(By.css('mat-card .button'));
      deleteButtons[0].triggerEventHandler('click', { stopPropagation() {} });

      tick();

      expect(deletedTodo).toEqual(todo);
    }));
  });
  describe('update', () => {
    it('should be raised when updateTodoCompletedStatus is called', fakeAsync(() => {
      const todo: Todo = {
        id: '1',
        user: '1',
        title: 'some title',
        description: 'some description',
        completed: true
      };

      component.todo = todo;
      fixture.detectChanges();

      let updatedTodo: Todo;
      component.update.subscribe(event => {
        updatedTodo = event;
      });

      component.updateTodoCompletedStatus(todo, true);

      tick();

      expect(updatedTodo).toEqual(todo);
    }));
    it('should be raised with the updated completed status', fakeAsync(() => {
      const todo: Todo = {
        id: '1',
        user: '1',
        title: 'some title',
        description: 'some description',
        completed: true
      };

      component.todo = todo;
      fixture.detectChanges();

      let updatedTodo: Todo;
      component.update.subscribe(event => {
        updatedTodo = event;
      });

      component.updateTodoCompletedStatus(todo, false);

      tick();

      expect(updatedTodo.completed).not.toEqual(todo.completed);
    }));
  });
});
