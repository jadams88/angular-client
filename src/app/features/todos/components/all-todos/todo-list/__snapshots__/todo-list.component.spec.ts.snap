// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListComponent  todo should displays the todo details when set 1`] = `
<div
  id="root3"
  ng-version="8.2.2"
>
  <mat-card>
    <mat-card-subtitle>
      some title
    </mat-card-subtitle>
    <mat-card-content
      fxlayout="row"
      fxlayoutalign="space-between center"
    >
      <span>
        some description
      </span>
      <mat-slide-toggle>
        Completed
      </mat-slide-toggle>
    </mat-card-content>
    <mat-card-actions
      fxlayoutalign="end"
    >
      <mat-icon
        class="button"
      >
        delete
      </mat-icon>
    </mat-card-actions>
  </mat-card>
</div>
`;
